### ISSUE TITLE
EXCEPTION_ACCESS_VIOLATION

### ISSUE THREAD

**Issue by ahmedoashery** [upx/upx#845]
after manual unpacking upx packed file i got this exception
uploaded full files 
any help please to unpack it
[Uploading upx_packed.zip…]()

**Comment by jreiser** [comment]
@ahmedoashery:   That supposed "Uploading upx_packed.zip" is not a .zip file, but rather an .html file. Please "Paste, drop, or click to add files" of the actual .zip file, not a .html file, and not a .html wrapper that points to a cloud storage download.  After some blank lines, then the file that **was** uploaded begins
```
!DOCTYPE html>
<html
  lang="en"
  
  data-color-mode="light" data-light-theme="light" data-dark-theme="dark"
  data-a11y-animated-images="system" data-a11y-link-underlines="true"
  >
    
  

  <head>
    <meta charset="utf-8">
  <link rel="dns-prefetch" href="https://github.githubassets.com">
  <link rel="dns-prefetch" href="https://avatars.githubusercontent.com">
```

Please follow our "New issue" template, which prompts for you to supply the hardware architecture and operating system of both the host and the target.

**Comment by ahmedoashery** [comment]
Sorry, 
here is the packed file

[Uploading UPX_Packed.zip…]()

**Comment by ahmedoashery** [comment]
[Uploading UPX_Packed.zip…]()

**Comment by ahmedoashery** [comment]
[Uploading UPX_Packed.zip…]()

**Comment by ahmedoashery** [comment]
[Program.zip](https://github.com/user-attachments/files/16816111/Program.zip)

**Comment by jreiser** [comment]
```
$ unzip -l Program.zip
Archive:  Program.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
 11859456  03-25-2022 09:01   ChilkatDelphi32.dll
     5490  07-08-2023 11:26   Sahl
  6834888  07-08-2023 11:26   UPX_Packed.exe
---------                     -------
 18699834                     3 files

$ sed 3q <Sahl
YMZYxaodu7plD4rLcHtDdEOBmmGx5weQESVg99axjBdg2MxxuUM5uMib/7IexeTIKQlmkq5UDFh3Fj7XdbXhfQ==
vjPYruEfQEcXEhzb65kFi221uMgcGBzrz4Ie2nSL9CHHWq/XVHL3iPYlbZ3ZfLYY7uLcUUqANUePfQ1w9ZP+JA==
zaz4sq0644rxRJKDcCOvx0X00XzlcrYp1Ft1nUOZLxcsyY/SY0UnvO8nRfv2zlCFyTMwuaQgvm/on6WBRbgzaA==
```

What is the `Sahl` file? The first 3 lines can be seen above.  It looks like a base64 encoding of some encrypted data.

The three-character string `UPX` appears only once in the file` UPX_Packed.exe`, but `UPX` appears at least 4 times in the output of any actual UPX compression.  The file does not contain any other small data "signatures" that UPX usually produces.  The sizes are also suspect:
```
-rw-r--r--. 1 user group 11859456 Mar 25  2022 ChilkatDelphi32.dll
-rw-r--r--. 1 user group  4518912 Mar 25  2022 packed-upx-1.exe
-rw-r--r--. 1 user group  3798016 Mar 25  2022 packed-upx-7.exe
-rw-r--r--. 1 user group 10843954 Aug 30 05:39 Program.zip
-rw-r--r--. 1 user group  6834888 Jul  8  2023 UPX_Packed.exe
```

That table of sizes shows that UPX at usual compression strength of `-7`, or at  least strength of `-1`, produces compressed files that are much shorter than `UPX_Packed.exe`.

It seems that you are confused.  Please confer with another human user of UPX (preferably in the same timezone) in order to clarify what is going on.

**Comment by ahmedoashery** [comment]
UPX_Packed.exe 
is upx packed file , which i wanna unpack it, but it says modified and protected  when try to unpack it using "upx -d"
Sahl is needed to start exe file

**Comment by jreiser** [comment]
Yes, UPX does not recognize UPX_Packed.exe as any possible output from UPX.  The file also does not look like any common modification that is intended to obscure packing by UPX, such as XOR of every byte with a constant.  As noted above, the size of the file UPX_Packed.exe bears no apparent relation to the compression by UPX of any known file.

UPX cannot provide any more information.  Can you?

**Comment by bitraid** [comment]
This file has the "UPX" strings replaced. You can restore them, and be able to unpack it  with `-d`.
[UPX_Packed-fixed.zip](https://github.com/user-attachments/files/16820844/UPX_Packed-fixed.zip)

**Comment by tansy** [comment]
> This file has the "UPX" strings replaced.

It is.
As far as I remember from similar case, it's against the licence to modify it.


> UPX_Packed-fixed.zip

No need to attach huge binaries to a post, especially it's ethically questionable file.
If you have to, add hex or binary patch like this:
[UPX_Packed-fixed.exe.hxd.diff.gz](https://github.com/user-attachments/files/16824078/UPX_Packed-fixed.exe.hxd.diff.gz)

**Comment by ahmedoashery** [comment]
> This file has the "UPX" strings replaced. You can restore them, and be able to unpack it with `-d`. [UPX_Packed-fixed.zip](https://github.com/user-attachments/files/16820844/UPX_Packed-fixed.zip)

i did this but but application does not work

**Comment by ahmedoashery** [comment]
> Yes, UPX does not recognize UPX_Packed.exe as any possible output from UPX. The file also does not look like any common modification that is intended to obscure packing by UPX, such as XOR of every byte with a constant. As noted above, the size of the file UPX_Packed.exe bears no apparent relation to the compression by UPX of any known file.
> 
> UPX cannot provide any more information. Can you?

![image](https://github.com/user-attachments/assets/2cbcd3a9-7ec9-4448-b3ab-c3935aec7a3f)
![image](https://github.com/user-attachments/assets/8ca211ca-2b9d-42e7-9920-b85dfe6c35ff)

**Comment by jreiser** [comment]
> > This file has the "UPX" strings replaced. You can restore them, and be able to unpack it with `-d`. [UPX_Packed-fixed.zip](https://github.com/user-attachments/files/16820844/UPX_Packed-fixed.zip)
> 
> i did this but but application does not work

@ahmedoashery   Please be specific.  What **exactly** did you do, step by step?  Give an actual copy+paste transcript of a terminal shell session (use copy+paste of text; do NOT use a screenshot), or an equivalent recipe in words.  Use actual file names; avoid pronouns.  What did you observe, in detail?  What did you expect, and what is the difference between observed and expected?  Explicitly pointing out the difference, more than just "look and see [visually compare]",  helps to increase communication and avoid glossing over details that could be important.

Does `UPX_Packed-fixed` run correctly?  What about UPX_Packed-fixed.unpacked [`upx -d -f -o UPX_Packed-fixed.unpacked UPX_Packed-fixed`]

Use a "system call recorder" or "system library call recorder" on each executable.  Show the results, and compare+contrast. In many cases, explicitly knowing how far execution progresses before failure gives good clues.

**Comment by tansy** [comment]
It decompresses, no problem. I don't know how could it fail for OP.

**Comment by jreiser** [comment]
> This file has the "UPX" strings replaced. You can restore them, and be able to unpack it with `-d`.

@bitraid:    Please share your methodology and tools to determine which bytes should be restored.

**Comment by jreiser** [comment]
> It decompresses, no problem. I don't know how could it fail for OP.

@tansy: Yes, `upx -d UPX_Packed-fixed.exe` finishes successfully with no complaints.  But the OP has not been specific about exactly which command line `does not work`. An app which refers its own code by using the filename whch invoked it often fails because the file is the compressed version.  It is a clue that an additional file `Sahl`, which contains encrypted data, must be present to invoke the app.  The `Sahl` file has no knowledge that the .exe was compressed later.

Also, successive alternating de-compression and re-compression gives inconsistent file sizes:
```
$ ls -l UPX_Packed*
-rw-r--r--. 1 user group  6834888 Jul  8  2023 UPX_Packed.exe
-rw-r--r--. 1 user group  6834888 Aug 30 20:38 UPX_Packed-fixed.exe
-rw-r--r--. 1 user group 24475848 Aug 30 20:38 UPX_Packed-fixed.exe.unpack
-rw-r--r--. 1 user group  7681224 Aug 30 20:38 UPX_Packed-fixed.exe.unpack.packed

$ upx --fileinfo UPX_Packed-fixed.exe UPX_Packed-fixed.exe.unpack.packed
UPX_Packed-fixed.exe [i386-win32.pe, win32/pe]
   6834888 bytes, compressed by UPX 13, method 8, level 7, filter 0x16/0x00
UPX_Packed-fixed.exe.unpack.packed [i386-win32.pe, win32/pe]
   7681224 bytes, compressed by UPX 13, method 8, level 7, filter 0x16/0x00
$ 
```
Thus `--fileinfo` says that compression parameters were the same, but the sizes differ by (7681224 - 6834888) = = 846336 bytes.
```

**Comment by bitraid** [comment]
@jreiser Nothing special - viewing the first bytes with a hex editor shows that the section names have been renamed, and the version string between the sections header and section 0 is modified.
Some useful tools are [PE-Bear](https://github.com/hasherezade/pe-bear), [XPE-Viewer](https://github.com/horsicq/XPEViewer) and maybe some advanced hex editor like *010 Editor*.

---

As for the file size difference, something strange is happening with upx:
Compressing `UPX_Packed-fixed.exe` with upx v4.2.4, gives different results on Linux (7681224), than on Windows (6834888). I don't have time to look more at this right now, but it's definitely something that needs to be figured out.

**Comment by bitraid** [comment]
OK, the size difference mystery is caused because we are testing upx binaries with different libraries (NRV vs UCL).

**Comment by tansy** [comment]
> successive alternating de-compression and re-compression gives inconsistent file sizes

You don't know which version of UPX and what filter/s were used to compress that file. Mind you the file was modified. Whether manually or with modified UPX version, no one knows.

```
$ wc -c UPX_Packed-fixed.exe* | sort -r

 24475848 UPX_Packed-fixed.exe.unp
  7447240 UPX_Packed-fixed.exe.unp.upx-3.96-ucl-7
  7422664 UPX_Packed-fixed.exe.unp.upx-3.96-ucl-8
  7252680 UPX_Packed-fixed.exe.unp.upx-3.96-ucl-9
  7152840 UPX_Packed-fixed.exe.unp.upx-4.2.4-5
  6980296 UPX_Packed-fixed.exe.unp.upx-4.2.4-6
  6980296 UPX_Packed-fixed.exe.unp.upx-3.96-6
  6834888 UPX_Packed-fixed.exe
  6608072 UPX_Packed-fixed.exe.unp.upx-4.2.4-7
  6608072 UPX_Packed-fixed.exe.unp.upx-3.96-9
  6608072 UPX_Packed-fixed.exe.unp.upx-3.96-7
  6472904 UPX_Packed-fixed.exe.unp.upx-4.2.4-8
  6472904 UPX_Packed-fixed.exe.unp.upx-3.96-8
  6387400 UPX_Packed-fixed.exe.unp.upx-4.2.4-9
```

OP can decompress it without a problem. If it doesn't run, it's not your nor our problem. Again, file was modified. It's not our job to fix hackers' job, whether it's screw up or purposeful act, we don;t know it either.

**Comment by jreiser** [comment]
> OK, the size difference mystery is caused because we are testing upx binaries with different libraries (NRV vs UCL).

@bitraid: Just for comparison, here is `--lzma -9`, which is about as small as possible among LZ methods that require a minimum of 3 characters in a match.  NRV and UCL require only a 2-character minimum (which should be an advantage), but lzma specially encodes up to 4 "most recently matched distances" while NRV and UCL have only 1.  (lzma takes longer when compressing, and much longer than NRV or UCL when de-compressing.)

```
$ upx --lzma -9 -f -o UPX_Packed-fixed.unpack.lzma9.exe UPX_Packed-fixed.exe.unpack
        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
  24475848 ->   5812424   23.75%    win32/pe     UPX_Packed-fixed.unpack.lzma9.exe
```

**Comment by bitraid** [comment]
> You don't know which version of UPX and what filter/s were used to compress that file.

Of course he can. He already shown it:
> compressed by UPX 13, method 8, level 7, filter 0x16/0x00

The issue is that method 8 uses the NRV compression library for the prebuilt upx binary, whereas the binary that comes from the distro repo uses the UCL.

---

>Whether manually or with modified UPX version, no one knows.

It can easily be determined whether the decompression stub has been modified or just the text strings.

---

> OP can decompress it without a problem. If it doesn't run, it's not your nor our problem. Again, file was modified. It's not our job to fix hackers' job, whether it's screw up or purposeful act, we don;t know it either.

One thing that stands out, is that  the file is digitally signed (after the compression/modification). That means that patching/unpacking/repacking will render the signature invalid, and the program could choose to terminate if this happens.

**Comment by bitraid** [comment]
> > OK, the size difference mystery is caused because we are testing upx binaries with different libraries (NRV vs UCL).
> 
> @bitraid: Just for comparison, here is `--lzma -9`, which is about as small as possible among LZ methods that require a minimum of 3 characters in a match. NRV and UCL require only a 2-character minimum (which should be an advantage), but lzma specially encodes up to 4 "most recently matched distances" while NRV and UCL have only 1. (lzma takes longer when compressing, and much longer than NRV or UCL when de-compressing.)
> 
> ```
> $ upx --lzma -9 -f -o UPX_Packed-fixed.unpack.lzma9.exe UPX_Packed-fixed.exe.unpack
>         File size         Ratio      Format      Name
>    --------------------   ------   -----------   -----------
>   24475848 ->   5812424   23.75%    win32/pe     UPX_Packed-fixed.unpack.lzma9.exe
> ```

But for non-lzma, make sure you are testing using upx binary from https://github.com/upx/upx/releases rather that the one that comes with the disto.

**Comment by tansy** [comment]
> >  You don't know which version of UPX and what filter/s were used to compress that file.
>
> Of course he can. He already shown it:
>
> >  compressed by UPX 13, method 8, level 7, filter 0x16/0x00

Not exactly.

~~~
$ upx-3.95 -o gzip.upx-3.95 gzip
$ upx-3.96 -o gzip.upx-3.96 gzip
$ upx --fileinfo gzip.upx-3.95 gzip.upx-3.96
gzip.upx-3.95 [i386-linux.elf, linux/i386]
     49644 bytes, compressed by UPX 13, method 2, level 8, filter 0x49/0x04
gzip.upx-3.96 [i386-linux.elf, linux/i386]
     49708 bytes, compressed by UPX 13, method 2, level 8, filter 0x49/0x04
~~~

One will not be able to figure out exact version, only that it was v3.xx.
Other than that, it's true.

I wonder, why does it store all those parameters, when they are not required for decompression.

**Comment by jreiser** [comment]
> I wonder, why does it store all those parameters, when they are not required for decompression.

Although not required for online de-compression when actually invoking the application, they are required in order to make it easy for offline de-compression (`upx -d`) and to make it easier for documentation and debugging.

**Comment by jreiser** [comment]
Closing. The compressed output was modified by some other actor.
